### PR TITLE
User documentation: Generate docs from macros, make`DocumentationBuilder::build` infallable

### PR DESCRIPTION
**Pull request by comphead**
## Which issue does this PR close?

Closes #.

## What changes are included in this PR?
- [x] Moved Document generation framework to separate crate
- [x] Created procedural macros crate and macros to generate user docs from rust attributes
- [x] Migrated UDF `to_date` as an example

### CODE REVIEW

**Comment by comphead** [comment]
@Omega359 @alamb I tried to play with custom attributes to wrap up the documentation on top of the what @Omega359 already built. I'm experimenting with just 2 fields(description and examples) and the attribute has to be placed on top of the struct.

File: Cargo.toml
Comment (comphead) [inline Cargo.toml:52]: pre macros needed to have Documentation structure already precreated

File: datafusion/functions/src/math/log.rs
Comment (comphead) [inline datafusion/functions/src/math/log.rs:41]: here we put the documentation

File: datafusion/functions/src/math/log.rs
Comment (comphead) [inline datafusion/functions/src/math/log.rs:476]: and the test shows it applied

**Comment by comphead** [comment]
WDYT should we move in this direction?

File: datafusion/functions/src/math/log.rs
Comment (alamb) [reply datafusion/functions/src/math/log.rs:41]: This looks quite cool ❤️ 

One thing I would be interested in seeing is if/how a multi line description and example would look. I have had trouble with multi-line macro / docs in the past

**Comment by Omega359** [comment]
I find this intriguing and technically very interesting! I do have a few questions:

- What are you goals with this? I am assuming cleaner looking/easier to create docs. If so I would wonder how it would help in cases were there is significant amounts of docs vs boilerplate such as in the [to_date](https://github.com/apache/datafusion/blob/3d347c99b32715690453add7fca1d76a3f8bbb01/datafusion/functions/src/datetime/to_date.rs#L84) udf.
- Can it be made to handle n arguments?
- Is it possible to generate rustdoc with this approach?
- I noticed that this approach creates a Documentation object on every call. It's obviously pretty minor but it would be nice if it could be static.

To me in general while I find this quite interesting I wonder if the benefit outweighs the added complexity of the code generation. To be clear I am absolutely not against this approach - I just have a few concerns.

**Comment by comphead** [comment]
@alamb @Omega359 please have a look on real example `to_date` (I still need to include argements to be called with the builder). As you can see it is the same approach as before, the only difference the documentation is set by attributes rather than code, but it generates documentation builders as before

File: datafusion/pre-macros/src/lib.rs
Comment (comphead) [inline datafusion/pre-macros/src/lib.rs:19]: I copied all Documentation structures here, added Test so as not to break everything

File: Cargo.toml
Comment (alamb) [reply Cargo.toml:52]: if we go with this approach, maybe we can put the doc structure in a crate named specifically such as  `datafusion/documentation` rather than something generic like `pre-macro`s

File: datafusion/functions/src/datetime/to_date.rs
Comment (alamb) [inline datafusion/functions/src/datetime/to_date.rs:33]: this is very cool

@Omega359 what do you think?

File: datafusion/macros/Cargo.toml
Comment (alamb) [inline datafusion/macros/Cargo.toml:20]:
```suggestion
description = "Procedural macros for DataFusion query engine"
```

**Review by alamb — COMMENTED** [review]
I think this looks great personally

As long as the macro has enough documentation itself, I think this would help reduce a bunch of the boiler plate (though perhaps not all)

**Comment by comphead** [comment]
Thanks @alamb I'll consider the changes, @Omega359 is it good to have a documentation method on struct level instead of impl?

**Comment by Omega359** [comment]
> @Omega359 is it good to have a `documentation` method on struct level instead of impl?

I can't think of a reason why not off the top of my head. I'm hoping to have time to look over your latest updates in this PR later tonight

File: datafusion/functions/src/datetime/to_date.rs
Comment (Omega359) [inline datafusion/functions/src/datetime/to_date.rs:34]: I worry about this since it's almost certainly going to drift at some point in the future via typos, etc. Is it possible to reference the const's in scalar_doc_sections/window_doc_sections/aggregate_doc_sections ?

File: datafusion/functions/src/datetime/to_date.rs
Comment (Omega359) [inline datafusion/functions/src/datetime/to_date.rs:57]: This is likely better as `expression_type` -> `prefix`

File: datafusion/macros/src/lib.rs
Comment (Omega359) [inline datafusion/macros/src/lib.rs:33]: might be missing related udfs here.

File: datafusion/functions/src/datetime/to_date.rs
Comment (Omega359) [reply datafusion/functions/src/datetime/to_date.rs:33]: I dumped the example #[udf_doc ...] into my editor to do a direct comparison to the existing solution and to be honest it definitely is nicer to visually parse. 

There are situations currently where this approach will not work, at least as it currently seems to be implemented. For example:  [bit_and_or_xor](https://github.com/apache/datafusion/blob/main/datafusion/functions-aggregate/src/bit_and_or_xor.rs), some of the regex udfs, etc.

**Comment by comphead** [comment]
Thanks @Omega359 there is bunch of things needed to be added, its not a final PR. I'll add some minor missing pieces and the PR will be ready for the review, most likely next week. Because I need to make a cleanup, remove test structures, polish crates and everything.

File: datafusion/functions/src/datetime/to_date.rs
Comment (comphead) [reply datafusion/functions/src/datetime/to_date.rs:34]: I'm not sure if enum/constant can be used directly within the attribute, however there is a possibility to check the value in macro. So if the value is not satisfied to existing enum variants the macro can throw an exception with explaining what section names are supported

**Comment by comphead** [comment]
I'm closer to this

**Comment by comphead** [comment]
We should also take care on handing things like https://github.com/apache/datafusion/pull/13367

File: Cargo.toml
Comment (comphead) [reply Cargo.toml:56]: Named it `doc-gen` as `documentation` sort of misleading, users gonna think this is a crate user documentation

Comment (alamb) [reply Cargo.toml:56]: proc-macros would also be a standard rust name for this kind of hting

Comment (alamb) [reply Cargo.toml:56]: We don't have any custom procedural macros, but I was thinking we could / should probably vendor (copy) the `recursive` one to reduce our dependency chain

Comment (comphead) [reply Cargo.toml]: I'm not quite following.... you mean copy `stacker/recurisive` macros introduced recently but this change not related to this PR?

File: datafusion/functions/src/datetime/to_date.rs
Comment (comphead) [inline datafusion/functions/src/datetime/to_date.rs:153]: I dont have better solutions here for now, so every impl have to copy this line because its not easy to inject/override a method into trait impl, I'll experiment more with this, but for now it is what it is

**Comment by comphead** [comment]
@alamb @Omega359 the PR is ready for review

File: datafusion/doc-gen/src/lib.rs
Comment (alamb) [inline datafusion/doc-gen/src/lib.rs:180]: Perhaps it is worth documenting that this will panic if doc section, description, or syntax example is missing

File: datafusion/doc-gen/Cargo.toml
Comment (alamb) [inline datafusion/doc-gen/Cargo.toml:19]: Since this holds `Documentation`, it seems like a name like `datafusion-doc` might be a name with more clarity

It would also likely help to explain why this needs to be in its own crate (so the macros can use it?)

File: datafusion/doc-gen/src/lib.rs
Comment (alamb) [inline datafusion/doc-gen/src/lib.rs:29]: I think leaving information about markdown is still valuable

File: datafusion/expr/src/lib.rs
Comment (alamb) [inline datafusion/expr/src/lib.rs:68]: 👍

File: datafusion/functions/src/datetime/to_date.rs
Comment (alamb) [reply datafusion/functions/src/datetime/to_date.rs:33]: SInce we auto generate the documentation, part of the PR review would be to look at the generated pages. I think we might be able to catch typos at that stage too

File: datafusion/functions/src/datetime/to_date.rs
Comment (alamb) [inline datafusion/functions/src/datetime/to_date.rs:41]: this is pretty fancy

File: datafusion/functions/src/datetime/to_date.rs
Comment (alamb) [reply datafusion/functions/src/datetime/to_date.rs:153]: I think as long as it is well documented, this boiler plate is ok

File: datafusion/macros/src/lib.rs
Comment (alamb) [inline datafusion/macros/src/lib.rs:24]: I think this macro needs user facing documentation (I realize that is quite meta 😆 ). Specifically a 
1. Introduction of what it does
2. A list of the optional arguments (`doc_section_include`, `doc_section_lbl`, etc) and examples
3. That it creates a standard `impl ... ` with a `doc_function`

It would also be really neat to make a doc example of it in use, if possible

File: datafusion/functions/src/datetime/to_date.rs
Comment (alamb) [reply datafusion/functions/src/datetime/to_date.rs:153]: Maybe we could call the function something like `macro_doc` or `derived_documentation` to make it easier to find (aka so someone who runs across it can easily `grep` and find the macro definition)

**Review by alamb — COMMENTED** [review]
I think this looks really neat in my opinion -- thank you @comphead 

Let's wait to see what @Omega359  has to say too, but I would be inclined to merge it in (as it is an optional feature) once we have a bit more documentation

File: datafusion/doc-gen/src/lib.rs
Comment (comphead) [reply datafusion/doc-gen/src/lib.rs:180]: Done. I was thinking to convert the builder to support constructor `new` with mandatory arguments so we dont need having this panics, I'll do it in follow up PR

File: datafusion/doc-gen/Cargo.toml
Comment (comphead) [reply datafusion/doc-gen/Cargo.toml:19]: Yes, procedural macros crates are not very flexible and requires all structures to be predefined before the macros is used. Renaming it

**Comment by comphead** [comment]
Done, 

> Maybe we could call the function something like macro_doc or derived_documentation to make it easier to find (aka so someone who runs across it can easily grep and find the macro definition)

Good idea, maybe we can name it as `user_documentation`. I'd prefer to leave it to the following PR though, because it will affect dozens of files

**Comment by comphead** [comment]
Another thing comes to my mind, we can automate transforming current API documentation approach to attribute based one wherever possible. There are more complicated options with macros, we can leave them with API approach for now

File: Cargo.toml
Comment (alamb) [inline Cargo.toml:105]: looks like the directory name is still `doc` when calling it `doc` would be more consistent with the others

File: datafusion/doc-gen/src/lib.rs
Comment (alamb) [reply datafusion/doc-gen/src/lib.rs:197]: I like that idea

**Review by alamb — APPROVED** [review]
Thank you @comphead  -- I think this looks really neat. 

Maybe we can wait a day or two before merging to let @Omega359 have a chance to comment if he would like

**Comment by Omega359** [comment]
> Let's wait to see what @Omega359 has to say too, but I would be inclined to merge it in (as it is an optional feature) once we have a bit more documentation

I'll hopefully be able to take a good look at this tonight 🤞

**Comment by Omega359** [comment]
Sorry, I've been busy lately. I'm hoping to get to reviewing this PR this week

File: datafusion/macros/src/lib.rs
Comment (Omega359) [inline datafusion/macros/src/lib.rs:23]:
```suggestion
/// This procedural macro is intended to parse a rust custom attribute and create user documentation
```

File: datafusion/macros/src/lib.rs
Comment (Omega359) [inline datafusion/macros/src/lib.rs:25]:
```suggestion
/// retrieved from the `documentation()` method
```

File: datafusion/macros/src/lib.rs
Comment (Omega359) [reply datafusion/macros/src/lib.rs:95]: Still missing related udfs.

File: datafusion/macros/src/lib.rs
Comment (Omega359) [inline datafusion/macros/src/lib.rs:95]: Missing alternative_syntax as well.

**Comment by comphead** [comment]
Thanks @Omega359 I will be adding the udfs and alternate syntax in follow up PR. I'll find the UDFs supporting both `related_udfs` and `alternative_syntax` and make these changes simultaneously along with some other things. Preferring the follow up PR as this one is already too large to review

**Comment by Omega359** [comment]
I'm fine with that @comphead - we can file followup issues to add the alternative syntax, etc

**Pull request merged**
